samwise.js

```diff
@@ -1,44 +1,47 @@
 /**
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?